src/app/connection-service.service.ts

```diff
@@ -15,38 +51,155 @@ export class ConnectionService {
+            ? loginSubjectOrObserveHandshake
+            : ConnectionService.loginBySessionTokenSubject;
+        ConnectionService.addConnection(connection, subscriber);
+        connection.pipe(this.rxjsSkip(loginSubjectOrObserveHandshake ? 0 : 2)).subscribe({
+            next: (message: Message) => subscriber.handleMessages(message),
+            complete: () => subscriber.handleComplete(),
+            error: (error: any) => subscriber.handleError(error)
+        });
+        connection.pipe(this.rxjsTake(2)).subscribe({
+            next: (message: Message) => this.handleHandshakeMessages(
+                message, {
+                    service: this,
+                    connection: connection,
+                    subscriber: subscriber,
+                    /* use either credentials from the subscriber or the local session token: */
+                    loginSubject: loginSubject,
+                    isPrimary: isPrimary == true
+                }
+            )
+            // ,complete: () => { console.log('handshake completed for', subscriber.componentID); }
+        });
+        console.groupEnd();
+    }
+    
+    // Handle the first two messages from a new connection, it should be a HelloMessage and a WelcomeMessage
+    // If the session token is not set yet, assume we receive it after successfull logon by 
+    // LoginComponent and send the ses_token through the sessionTokenSubject
+    handleHandshakeMessages(
+        message: Message,
+        that?: {
+            service: ConnectionService,
+            connection: rxws.WebSocketSubject<Message>,
+            subscriber: ConnectedComponent,
+            loginSubject: LoginSubject,
+            isPrimary: boolean
+        }
+    ) {
+        console.groupCollapsed('handle handshake: ', message.type, 'to', that?.subscriber.componentID);
+        console.log( message); console.log('that:', that);
+        console.groupEnd();
+        if (message instanceof HelloMessage) {
+            if (that) {
+                console.log(that.subscriber.componentID, 'awaits credentials')
+                that.subscriber.setToken(message.token);
+                that.loginSubject.pipe(rxjs.take(1)).subscribe(
+                    (credentials: LoginCredentials) => {
+                        console.log(that.subscriber.componentID, 'got credentials:', credentials);
+                        that.service.sendMessage(
+                            new LoginMessage(credentials, message.token, that.isPrimary),
+                            that.subscriber.componentID
+                        );
+                    }
+                );
+                // no need to unsubscribe because take(1) implicates that
+            }
+        } else if (message instanceof WelcomeMessage) {
+            // if the session token is not set yet provide it for other connections
+            if (message.ses_token && ! ConnectionService._sessionToken ) {
```

Comment:
Handle case if session token is already set on the service, but backend sends a different session token.

Reply:
'throw' added